lib_eio_linux/eio_linux.ml

```diff
@@ -1162,11 +1162,90 @@ let net = object
   method getnameinfo = Eio_unix.getnameinfo
 end
 
+let rec wait_for_process flags pid = 
+  try Unix.waitpid flags pid with
+  | Unix.Unix_error (Unix.EINTR, _, _) -> wait_for_process flags pid
+
+external sig_to_host_sig : int -> int = "caml_signal_to_posix_signal"
+
+let pid_to_process close pid = object
+  inherit Eio.Process.t
+  method pid = pid
+  val mutable needs_close = true
+  method needs_close = needs_close
+  method set_needs_close b = needs_close <- b
+  method status =
+    let p, r = Promise.create () in
+    let run () =
+      let v = wait_for_process [] pid in 
+      if needs_close then (close (); needs_close <- false);
+      match v with
+      | _, WEXITED i -> Promise.resolve r  (Eio.Process.Exited (sig_to_host_sig i))
+      | _, WSIGNALED i -> Promise.resolve r  (Eio.Process.Signaled (sig_to_host_sig i))
+      | _, WSTOPPED i -> Promise.resolve r  (Eio.Process.Stopped (sig_to_host_sig i))
+    in
+    Eio_unix.run_in_systhread run;
+    p
+  
+  method stop = 
+    Unix.kill pid Sys.sigkill
+end
+
+let get_fd_or_err flow =
+  match get_fd_opt flow with
+  | Some fd -> 
+    Unix.dup ~cloexec:false (FD.to_unix `Peek fd)
+  | None -> failwith "Currently only flows backed by file descriptors are supported!"
+
+let process = object
+  inherit Eio.Process.mgr
+
+  (* TODO: Is Sys.chdir cwd domain-safe, ? *)
+  method spawn ~sw ?cwd ~stdin ~stdout ~stderr cmd args =
+    let stdin = get_fd_or_err stdin in
+    let stdout = get_fd_or_err stdout in
+    let stderr = get_fd_or_err stderr in
+    let cwd = Option.map snd cwd in
+    let pid =
+      Option.iter Sys.chdir cwd;
+      Unix.create_process cmd
+        (Array.of_list args) 
+        stdin stdout stderr
+    in
+    let close () =
+      Unix.close stdin;
+      Unix.close stdout;
+      Unix.close stderr
+    in
+    let process = pid_to_process close pid in
+    let cleanup () =
+      (* Catch if the process is already finished when trying to stop it. *)
+      (try process#stop with Unix.Unix_error (Unix.ESRCH, _, _) -> ());
```

Comment:
Using `pidfd_open(2)` would likely solve these races, and would avoid the need for a systhread to do the waitpid.